File: framework/ui/src/io/intino/alexandria/ui/displays/components/RangeSlider.java
package io.intino.alexandria.ui.displays.components;

import io.intino.alexandria.core.Box;
import io.intino.alexandria.schemas.RangeSelected;
import io.intino.alexandria.schemas.RangeValue;
import io.intino.alexandria.schemas.Selected;
import io.intino.alexandria.ui.displays.components.slider.Ordinal;
import io.intino.alexandria.ui.displays.components.slider.Range;
import io.intino.alexandria.ui.displays.notifiers.RangeSliderNotifier;

public class RangeSlider<DN extends RangeSliderNotifier, B extends Box> extends AbstractRangeSlider<DN, B> {

	public RangeSlider(B box) {
		super(box);
		_value(0, 0);
	}

	@Override
	public void reset() {
		Range range = range();
		_value(range.min(), range.max());
	}

	protected RangeSlider<DN, B> _value(long from, long to) {
		return _value(new RangeValue().from(from).to(to));
	}

	protected RangeSlider<DN, B> _value(RangeValue value) {
		this.value = value;
		return this;
	}

	public RangeSlider<DN, B> range(long min, long max) {
		_range(min, max);
		notifier.refreshRange(rangeSchema());
		return this;
	}

	public String formattedValue(long value) {
		return format(value);
	}

	public void moved(RangeValue value) {
		notifier.refreshSelected(schemaOf(value));
		notifier.refreshToolbar(toolbarState());
	}

	public void update(RangeValue value) {
		value(value);
	}

	@Override
	String format(long value) {
		Ordinal ordinal = ordinal();
		if (ordinal == null) ordinal = defaultOrdinals().get(0);
		return ordinal != null ? ordinal.formatter(language()).format(value) : String.valueOf(value);
	}

	@Override
	void updateRange() {
	}

	@Override
	<T extends Selected> void refreshSelected(T schema) {
		notifier.refreshSelected((RangeSelected) schema);
	}

	@Override
	protected void _value(Object value) {
		this.value = value;
	}

	@Override
	protected void _range(long min, long max) {
		this.range = new Range().min(min).max(max);
		RangeValue value = value();
		if (value.from() < min) value.from(min);
		if (value.to() < min) value.to(min);
		if (value.from() > max) value.from(max);
		if (value.to() > max) value.to(max);
		_value(value);
	}

	@Override
	public void previous() {
		RangeValue values = (RangeValue)value;
		values.from(values.from()-1);
		values.to(values.to()-1);
		value(values);
	}

	@Override
	public void next() {
		RangeValue values = (RangeValue)value;
		values.from(values.from()+1);
		values.to(values.to()+1);
		value(values);
	}

	@Override
	protected boolean canPrevious() {
		return ((RangeValue)value()).from() > range.min;
	}

	@Override
	protected boolean canNext() {
		return ((RangeValue)value()).to() < range.max;
	}

	@Override
	protected void nextValue() {
		RangeValue value = value();
		if (value.to() >= range.max) {
			if (canLoop()) {
				value.to(value.to()-value.from()-1);
				value.from(range.min()-1);
				value(value);
			}
			else return;
		}
		value.from(value.from()+1);
		value.to(value.to()+1);
		value(value);
	}

	@Override
	protected boolean checkRange(Object value) {
		RangeValue values = (RangeValue) value;
		if (values.to() > range.max) return false;
		if (values.from() < range.min) return false;
		return true;
	}

	@Override
	protected RangeSelected schemaOf(Object value) {
		RangeValue values = (RangeValue)value;
		String formattedFrom = formattedValue(values.from());
		String formattedTo = formattedValue(values.to());
		return (RangeSelected) new RangeSelected().to(values.to()).formattedTo(formattedTo).value(values.from()).formattedValue(formattedFrom);
	}

}

File: framework/ui-elements/src/displays/components/RangeSlider.js
import React from "react";
import { TextField } from '@material-ui/core';
import { withStyles } from '@material-ui/core/styles';
import { withSnackbar } from 'notistack';
import AbstractRangeSlider from "../../../gen/displays/components/AbstractRangeSlider";
import RangeSliderNotifier from "../../../gen/displays/notifiers/RangeSliderNotifier";
import RangeSliderRequester from "../../../gen/displays/requesters/RangeSliderRequester";
import DisplayFactory from "alexandria-ui-elements/src/displays/DisplayFactory";
import BaseSlider from "./BaseSlider";
import Delayer from "../../util/Delayer";

const styles = theme => ({...BaseSlider.Styles(theme)});

class RangeSlider extends AbstractRangeSlider {

	constructor(props) {
		super(props);
		this.notifier = new RangeSliderNotifier(this);
		this.requester = new RangeSliderRequester(this);
	};

	render() {
		return this.renderComponent();
	};

	getValue = () => {
	    const selected = this.state.selected;
		return selected != null && selected.value !== undefined ? [selected.value, selected.to] : [0, 0];
	};

	getFormattedValue = () => {
	    const selected = this.state.selected;
		return selected != null && selected.value !== undefined ? selected.formattedValue + " " + this.translate("to") + " " + selected.formattedTo : 0 + " " + this.translate("to") + " " + 0;
	};

	renderValueEditor = () => {
		const range = this.state.range;
		return (
		    <React.Fragment>
                <TextField format={this.variant("body1")} type="number"
                           value={this.state.editorValue} onChange={this.handleValueEditorChange.bind(this)} onKeyPress={this.handleValueEditorKeyPress.bind(this)}
                           style={{width:'100%'}} autoFocus={true}
                           inputProps={{
                               min: range.min !== -1 ? range.min : undefined,
                               max: range.max !== -1 ? range.max : undefined,
                               step: 1
                           }}/>
                <TextField format={this.variant("body1")} type="number"
                           value={this.state.editorTo} onChange={this.handleToEditorChange.bind(this)} onKeyPress={this.handleValueEditorKeyPress.bind(this)}
                           style={{width:'100%',marginTop:'10px'}} autoFocus={true}
                           inputProps={{
                               min: range.min !== -1 ? range.min : undefined,
                               max: range.max !== -1 ? range.max : undefined,
                               step: 1
                           }}/>
            </React.Fragment>
		);
	};

	showValueDialog = () => {
	    this.setState({openValueDialog: true, editorValue: this.state.selected != null ? this.state.selected.value : this.state.range.min, editorTo: this.state.selected != null ? this.state.selected.to : this.state.range.min+1 });
	};

	handleToEditorChange = (e) => {
	    this.setState({editorTo: e.target.value});
	};

	handleValueChange = () => {
		this.requester.update({from: this.state.editorValue, to: this.state.editorTo});
		this.setState({openValueDialog: false});
	};

	handleMoved = (e, newValue) => {
	    const minDistance = this.props.minimumDistance;
	    const activeThumb = this.state.selected != null && Math.abs(this.state.selected.value - newValue[0]) >= 1 ? 0 : 1;
	    let value = newValue;
        if (minDistance != null && (newValue[1] - newValue[0] < minDistance)) {
          if (activeThumb === 0) {
            const clamped = Math.min(newValue[0], 100 - minDistance);
            value = [clamped, clamped + minDistance];
          } else {
            const clamped = Math.max(newValue[1], minDistance);
            value = [clamped - minDistance, clamped];
          }
        }
		this.requester.moved({from: value[0], to: value[1]});
		this.setState({value});
	};

	handleChange = (e, value) => {
		Delayer.execute(this, () => this.requester.update({from: value[0], to: value[1]}));
		this.setState({value});
	};

	handleFormattedValue = (value, index) => {
	    const selected = this.state.selected;
	    if (selected == null) return value;
	    return selected.value == value ? selected.formattedValue : selected.formattedTo;
	};

}

export default withStyles(styles, { withTheme: true })(withSnackbar(RangeSlider));
DisplayFactory.register("RangeSlider", withStyles(styles, { withTheme: true })(withSnackbar(RangeSlider)));